Crash in minimizer

```
martin@mediaNUK:~/go/src/github.com/holiman/goevmlab/cmd/minimizer$ ./minimizer --geth ~/workspace/evm --nethermind ~/workspace/nethtest ~/go/src/github.com/holiman/goevmlab/cmd/generic-fuzz
er/save-0053/00000053-naivefuzz-1.json                                                                                                                                                        
INFO [12-03|12:44:58.468] Mutating gas                             value=4,000,000                                                                                                            
INFO [12-03|12:44:59.008] Change ok                                                                                                                                                           
INFO [12-03|12:44:59.008] Mutating gas                             value=2,000,000                                                                                                            
INFO [12-03|12:44:59.566] Change ok                                                                                                                                                           
INFO [12-03|12:44:59.566] Mutating gas                             value=1,000,000                                                                                                            
INFO [12-03|12:45:00.113] Change ok                                                                                                                                                           
INFO [12-03|12:45:00.114] Mutating gas                             value=500,000                                                                                                              
INFO [12-03|12:45:00.660] Change ok                                                                                                                                                           
INFO [12-03|12:45:00.660] Mutating gas                             value=250,000                                                                                                              
INFO [12-03|12:45:01.196] Change ok                                                                                                                                                           
INFO [12-03|12:45:01.197] Mutating gas                             value=125,000                                                                                                              
INFO [12-03|12:45:01.740] Change ok                                                                                                                                                           
INFO [12-03|12:45:01.740] Mutating gas                             value=62500                                                                                                                
INFO [12-03|12:45:02.274] Change ok                                                                                                                                                           
INFO [12-03|12:45:02.275] Mutating gas                             value=31250                                                                                                                
INFO [12-03|12:45:02.807] Change ok                                                                                                                                                           
INFO [12-03|12:45:02.807] Mutating gas                             value=15625                                                                                                                
panic: no stateroot found                                                                                                                                                                                                                                                                                                                                                                   
goroutine 1 [running]:                                                                                                                                                                        
main.startFuzzer.func2()                                                                                                                                                                      
        /home/martin/go/src/github.com/holiman/goevmlab/cmd/minimizer/main.go:158 +0x1ad                                                                                                      
main.startFuzzer.func3(0x3d09)                                                                                                                                                                
        /home/martin/go/src/github.com/holiman/goevmlab/cmd/minimizer/main.go:175 +0xd0                                                                                                       sort.Search(0x8defa0?, 0xc0004c58d0)                                                                                                                                                          
        /usr/local/go/src/sort/search.go:65 +0x48                                                                                                                                             
main.startFuzzer(0xc0003f2400)                                                                                                                                                                
        /home/martin/go/src/github.com/holiman/goevmlab/cmd/minimizer/main.go:172 +0x39c                                                                                                      
github.com/urfave/cli/v2.(*Command).Run(0xc0003fe780, 0xc0003f2400, {0xc000024180, 0x6, 0x6})                                                                                                 
        /home/martin/go/pkg/mod/github.com/urfave/cli/v2@v2.23.5/command.go:271 +0xaab                                                                                                        
github.com/urfave/cli/v2.(*App).RunContext(0xc0001fe5a0, {0xa4ea10?, 0xc0000260a8}, {0xc000024180, 0x6, 0x6})                                                                                         /home/martin/go/pkg/mod/github.com/urfave/cli/v2@v2.23.5/app.go:329 +0x665                                                                                                            
github.com/urfave/cli/v2.(*App).Run(...)                                                                                                                                                      
        /home/martin/go/pkg/mod/github.com/urfave/cli/v2@v2.23.5/app.go:306                                                                                                                   
main.main()                                                                                                                                                                                   
        /home/martin/go/src/github.com/holiman/goevmlab/cmd/minimizer/main.go:57 +0x99                                                                                                        
martin@mediaNUK:~/go/src/github.com/holiman/goevmlab/cm
```

--- a/cmd/minimizer/main.go
+++ b/cmd/minimizer/main.go
@@ -117,7 +117,17 @@ func startFuzzer(c *cli.Context) error {
 		testPath  = c.Args().First()
 		compareFn func(path string, c *cli.Context) (bool, error)
 	)
-	compareFn = common.RootsEqual
+	compareFn = func(path string, c *cli.Context) (bool, error) {
+		agree, err := common.RootsEqual(path, c)
+		// An error here might mean that e.g the gas was changed so that
+		// the intrinsic gas was made wrong, and this the tx is invalid.
+		// we can ignore that error, and report it as 'agree', triggering the
+		// revertal of whatever it was that caused it
+		if err != nil {
+			return true, nil
+		}
+		return agree, nil
+	}
 	if c.Bool(fullTraceFlag.Name) {
 		compareFn = func(path string, c *cli.Context) (bool, error) {
 			agree, err := common.RunTests([]string{path}, c)

--- a/common/utils.go
+++ b/common/utils.go
@@ -153,6 +153,7 @@ func RootsEqual(path string, c *cli.Context) (bool, error) {
 			root, _, err := vm.GetStateRoot(path)
 			roots[index] = root
 			errs[index] = err
+			vm.Close()
 			wg.Done()
 		}(i, vm)
 	}

--- a/evms/besu.go
+++ b/evms/besu.go
@@ -17,9 +17,15 @@
 package evms
 
 import (
+	"bufio"
+	"bytes"
+	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
 	"io"
+	"os"
 	"os/exec"
 	"strings"
 )
@@ -44,9 +50,9 @@ func (evm *BesuVM) Name() string {
 // RunStateTest implements the Evm interface
 func (evm *BesuVM) RunStateTest(path string, out io.Writer, speedTest bool) (string, error) {
 	var (
-		stdout io.ReadCloser
 		err    error
 		cmd    *exec.Cmd
+		stdout io.ReadCloser
 	)
 	if speedTest {
 		cmd = exec.Command(evm.path, "--nomemory", "state-test", path)
@@ -67,8 +73,7 @@ func (evm *BesuVM) RunStateTest(path string, out io.Writer, speedTest bool) (str
 	return cmd.String(), nil
 }
 
-func (vm *BesuVM) Close() {
-}
+func (vm *BesuVM) Close() {}
 
 func (vm *BesuVM) GetStateRoot(path string) (root, command string, err error) {
 	// Run without tracing
@@ -84,11 +89,70 @@ func (vm *BesuVM) GetStateRoot(path string) (root, command string, err error) {
 		root := string(data[start : start+2+64])
 		return root, cmd.String(), nil
 	}
-	return "", cmd.String(), errors.New("no stateroot found")
+	log.Error("Failed to find stateroot", "vm", vm.Name(), "cmd", cmd.String())
+	return "", cmd.String(), errors.New("besu: no stateroot found")
 }
 
 // feed reads from the reader, does some geth-specific filtering and
 // outputs items onto the channel
 func (evm *BesuVM) Copy(out io.Writer, input io.Reader) {
-	besuCopyUntilEnd(out, input)
+	evm.copyUntilEnd(out, input)
+}
+
+type besuStateRoot struct {
+	StateRoot string `json:"postHash"`
+}
+
+func (evm *BesuVM) copyUntilEnd(out io.Writer, input io.Reader) stateRoot {
+	var stateRoot stateRoot
+	scanner := bufio.NewScanner(input)
+	// We use a larger scanner buffer for besu: it does not have a way to
+	// disable 'returndata', which can become larger than fits into a default
+	// scanner buffer
+	buf := make([]byte, 16*1024*1024)
+	scanner.Buffer(buf, cap(buf))
+	for scanner.Scan() {
+		data := scanner.Bytes()
+		var elem logger.StructLog
+		// Besu sometimes report a negative refund
+		if i := bytes.Index(data, []byte(`"refund":-`)); i > 0 {
+			// we can just make it positive, it will be zeroed later
+			data[i+9] = byte(' ')
+		}
+		err := json.Unmarshal(data, &elem)
+		if err != nil {
+			fmt.Printf("besub err: %v, line\n\t%v\n", err, string(data))
+			continue
+		}
+		// If the output cannot be marshalled, all fields will be blanks.
+		// We can detect that through 'depth', which should never be less than 1
+		// for any actual opcode
+		if elem.Depth == 0 {
+			if stateRoot.StateRoot == "" {
+				var tempRoot besuStateRoot
+				if err := json.Unmarshal(data, &tempRoot); err == nil {
+					// Besu calls state root "postHash"
+					stateRoot.StateRoot = tempRoot.StateRoot
+				}
+			}
+			// If we have a stateroot, we're done
+			break
+		}
+		// When geth encounters end of code, it continues anyway, on a 'virtual' STOP.
+		// In order to handle that, we need to drop all STOP opcodes.
+		if elem.Op == 0x0 {
+			continue
+		}
+		RemoveUnsupportedElems(&elem)
+		jsondata, _ := json.Marshal(elem)
+		if _, err := out.Write(append(jsondata, '\n')); err != nil {
+			fmt.Fprintf(os.Stderr, "Error writing to out: %v\n", err)
+			return stateRoot
+		}
+	}
+	root, _ := json.Marshal(stateRoot)
+	if _, err := out.Write(append(root, '\n')); err != nil {
+		fmt.Fprintf(os.Stderr, "Error writing to out: %v\n", err)
+	}
+	return stateRoot
 }

--- a/evms/besubatcher.go
+++ b/evms/besubatcher.go
@@ -17,33 +17,28 @@
 package evms
 
 import (
-	"bufio"
-	"encoding/json"
 	"fmt"
 	"io"
-	"os"
 	"os/exec"
 	"sync"
-
-	"bytes"
-	"github.com/ethereum/go-ethereum/eth/tracers/logger"
 )
 
 // BesuBatchVM is s Evm-interface wrapper around the `evmtool` binary, based on Besu.
 // The BatchVM spins up one 'master' instance of the VM, and uses that to execute tests
 type BesuBatchVM struct {
-	path   string
+	BesuVM
 	cmd    *exec.Cmd // the 'master' process
 	stdout io.ReadCloser
 	stdin  io.WriteCloser
 	mu     sync.Mutex
-	name   string // in case multiple instances are used
 }
 
 func NewBesuBatchVM(path, name string) *BesuBatchVM {
 	return &BesuBatchVM{
-		path: path,
-		name: name,
+		BesuVM: BesuVM{
+			path: path,
+			name: name,
+		},
 	}
 }
 
@@ -82,7 +77,7 @@ func (evm *BesuBatchVM) RunStateTest(path string, out io.Writer, speedTest bool)
 	defer evm.mu.Unlock()
 	_, _ = evm.stdin.Write([]byte(fmt.Sprintf("%v\n", path)))
 	// copy everything for the _current_ statetest to the given writer
-	besuCopyUntilEnd(out, evm.stdout)
+	evm.copyUntilEnd(out, evm.stdout)
 	// release resources, handle error but ignore non-zero exit codes
 	return evm.cmd.String(), nil
 }
@@ -113,77 +108,12 @@ func (evm *BesuBatchVM) GetStateRoot(path string) (root, command string, err err
 	evm.mu.Lock()
 	defer evm.mu.Unlock()
 	_, _ = evm.stdin.Write([]byte(fmt.Sprintf("%v\n", path)))
-	sRoot := besuCopyUntilEnd(devNull{}, evm.stdout)
+	sRoot := evm.copyUntilEnd(devNull{}, evm.stdout)
 	return sRoot.StateRoot, evm.cmd.String(), nil
-
 }
 
 type devNull struct{}
 
 func (d devNull) Write(p []byte) (n int, err error) {
 	return len(p), nil
-}
-
-// Copy feed reads from the reader, does some client-specific filtering and
-// outputs BesuBatchVM onto the channel
-func (evm *BesuBatchVM) Copy(out io.Writer, input io.Reader) {
-	besuCopyUntilEnd(out, input)
-}
-
-type besuStateRoot struct {
-	StateRoot string `json:"postHash"`
-}
-
-func besuCopyUntilEnd(out io.Writer, input io.Reader) stateRoot {
-	var stateRoot stateRoot
-	scanner := bufio.NewScanner(input)
-	// We use a larger scanner buffer for besu: it does not have a way to
-	// disable 'returndata', which can become larger than fits into a default
-	// scanner buffer
-	buf := make([]byte, 16*1024*1024)
-	scanner.Buffer(buf, cap(buf))
-	for scanner.Scan() {
-		data := scanner.Bytes()
-		var elem logger.StructLog
-		// Besu sometimes report a negative refund
-		if i := bytes.Index(data, []byte(`"refund":-`)); i > 0 {
-			// we can just make it positive, it will be zeroed later
-			data[i+9] = byte(' ')
-		}
-		err := json.Unmarshal(data, &elem)
-		if err != nil {
-			fmt.Printf("besub err: %v, line\n\t%v\n", err, string(data))
-			continue
-		}
-		// If the output cannot be marshalled, all fields will be blanks.
-		// We can detect that through 'depth', which should never be less than 1
-		// for any actual opcode
-		if elem.Depth == 0 {
-			if stateRoot.StateRoot == "" {
-				var tempRoot besuStateRoot
-				if err := json.Unmarshal(data, &tempRoot); err == nil {
-					// Besu calls state root "postHash"
-					stateRoot.StateRoot = tempRoot.StateRoot
-				}
-			}
-			// If we have a stateroot, we're done
-			break
-		}
-		// When geth encounters end of code, it continues anyway, on a 'virtual' STOP.
-		// In order to handle that, we need to drop all STOP opcodes.
-		if elem.Op == 0x0 {
-			continue
-		}
-		RemoveUnsupportedElems(&elem)
-		jsondata, _ := json.Marshal(elem)
-		if _, err := out.Write(append(jsondata, '\n')); err != nil {
-			fmt.Fprintf(os.Stderr, "Error writing to out: %v\n", err)
-			return stateRoot
-		}
-	}
-	root, _ := json.Marshal(stateRoot)
-	if _, err := out.Write(append(root, '\n')); err != nil {
-		fmt.Fprintf(os.Stderr, "Error writing to out: %v\n", err)
-	}
-	return stateRoot
 }

--- a/evms/erigon.go
+++ b/evms/erigon.go
@@ -27,6 +27,7 @@ import (
 	"strings"
 
 	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
 )
 
 // ErigonVM is s Evm-interface wrapper around the eroigon `evm` binary
@@ -63,7 +64,8 @@ func (evm *ErigonVM) GetStateRoot(path string) (root, command string, err error)
 		root := fmt.Sprintf("0x%v", string(data[start+len("mismatch: got "):end]))
 		return root, cmd.String(), nil
 	}
-	return "", cmd.String(), errors.New("no stateroot found")
+	log.Error("Failed to find stateroot", "vm", evm.Name(), "cmd", cmd.String())
+	return "", cmd.String(), errors.New("besu: no stateroot found")
 }
 
 // RunStateTest implements the Evm interface

--- a/evms/geth.go
+++ b/evms/geth.go
@@ -27,6 +27,7 @@ import (
 	"strings"
 
 	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
 )
 
 // GethEVM is s Evm-interface wrapper around the `evm` binary, based on go-ethereum.
@@ -63,7 +64,8 @@ func (evm *GethEVM) GetStateRoot(path string) (root, command string, err error) 
 		root := fmt.Sprintf("0x%v", string(data[start+len("mismatch: got "):end]))
 		return root, cmd.String(), nil
 	}
-	return "", cmd.String(), errors.New("no stateroot found")
+	log.Error("Failed to find stateroot", "vm", evm.Name(), "cmd", cmd.String())
+	return "", cmd.String(), errors.New("geth: no stateroot found")
 }
 
 // RunStateTest implements the Evm interface

--- a/evms/nethermind.go
+++ b/evms/nethermind.go
@@ -59,7 +59,7 @@ func (evm *NethermindVM) GetStateRoot(path string) (root, command string, err er
 	marker := `{"stateRoot":"`
 	start := strings.Index(string(data), marker)
 	if start <= 0 {
-		log.Info("Nethermind error", "err", "no stateroot found", "command", cmd)
+		log.Error("Failed to find stateroot", "vm", evm.Name(), "cmd", cmd.String())
 		return "", cmd.String(), errors.New("nethermind: no stateroot found")
 	}
 	end := strings.Index(string(data)[start:], `"}`)
@@ -67,7 +67,8 @@ func (evm *NethermindVM) GetStateRoot(path string) (root, command string, err er
 		root := string(data[start+len(marker) : start+end])
 		return root, cmd.String(), nil
 	}
-	return "", cmd.String(), errors.New("no stateroot found")
+	log.Error("Failed to find stateroot", "vm", evm.Name(), "cmd", cmd.String())
+	return "", cmd.String(), errors.New("nethermind: no stateroot found")
 }
 
 // RunStateTest implements the Evm interface

--- a/evms/nethermind.go
+++ b/evms/nethermind.go
@@ -28,6 +28,7 @@ import (
 	"strings"
 
 	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
 )
 
 // NethermindVM is s Evm-interface wrapper around the `nethtest` binary, based on Nethermind.
@@ -55,11 +56,11 @@ func (evm *NethermindVM) GetStateRoot(path string) (root, command string, err er
 	if err != nil {
 		return "", cmd.String(), err
 	}
-	//fmt.Printf("cmd: '%v', output: %v\n", cmd.String(),string(data))
 	marker := `{"stateRoot":"`
 	start := strings.Index(string(data), marker)
 	if start <= 0 {
-		return "", cmd.String(), errors.New("no stateroot found")
+		log.Info("Nethermind error", "err", "no stateroot found", "command", cmd)
+		return "", cmd.String(), errors.New("nethermind: no stateroot found")
 	}
 	end := strings.Index(string(data)[start:], `"}`)
 	if start > 0 && end > 0 {

--- a/evms/testdata/README.md
+++ b/evms/testdata/README.md
@@ -1,6 +1,23 @@
+## Testdata output
+
+This folder contains 
+ - statetests, 
+ - For each statetest: 
+   - output from vms, `stdout` and `stderr` output
+
+The statetests have been chosen because they trigger some quirk in a vm, e.g a statetest may 
+trigger a negative refund in besu. If we want to change the besu-shim at some later point, 
+when said bug has been fixed, we need to regenerate the outputs and check if the 
+tests passes. 
+
 ## Command to generate these
 
-# First set the binaries to use
+The script below, after setting the binaries to use, should recreate the outputs 
+in `traces` based on the inputs in `cases`. 
+
+```bash
+#!/bin/bash
+
 # evm="/home/martin/workspace/evm"
 # nethtest="/home/martin/workspace/nethtest"
 # besuvm="/home/martin/workspace/besu-vm"
@@ -8,7 +25,7 @@
 
 ### Geth
 
-if [[ -n "$geth" ]]; then
+if [[ -n "geth" ]]; then
     echo "geth"
     for i in ./cases/*.json; do
         $evm --json --nomemory --noreturndata statetest $i 2>/dev/null 1>./traces/$i.geth.stdout.txt
@@ -47,3 +64,4 @@ if [[ -n "$erigonvm" ]]; then
         $erigonvm --json --nomemory --noreturndata statetest $i 1>/dev/null 2>./traces/$i.erigon.stderr.txt
     done
 fi
+```

